week-07-rails-continued/day-03/module-01/readme.md

```diff
@@ -1 +1,404 @@
+  english = Course.create(name: "English")
+  french = Course.create(name: "French")
+  science = Course.create(name: "Science")
+
+  # associate our model instances
+  sally.courses << algebra
+  # ^ same as:
+  # sally.courses.push(algebra)
+  sally.courses << french
+
+  fred.courses << science
+  fred.courses << english
+  fred.courses << french
+
+  # here's a little trick: use an array to associate multiple courses with a student in just one line of code
+  alice.courses << [english, algebra]
+  ```
+
+  **Note:** Because we've used `through`, we can create our associations in the same way we do for a 1:N association (`<<`).
+
+3. Still in the Rails console, test your data to make sure your associations worked:
+
+  ```ruby
+  sally.courses.map(&:name)
+  # => ["Algebra", "French"]
+
+  fred.courses.map(&:name)
+  # => ["Science", "English", "French"]
+
+  alice.courses.map(&:name)
+  # => ["English", "Algebra"]
+  ```
+
+## Challenges, Part 2: Many-To-Many
+
+Head over to the [Many-To-Many Challenges](many-to-many-challenges.md) and work together in pairs.
+
+
+## Stretch Challenge: Self-Referencing Associations
+
+Lots of real-world apps create associations between items that are the same type of resource.  Read (or reread) <a href="http://guides.rubyonrails.org/association_basics.html#self-joins" target="_blank">the "self joins" section of the Associations Basics Rails Guide</a>, and try to create a self-referencing association in your `practice_associations` app. (Classic use cases are friends and following, where both related resources would be users.)
+
+## Migration Workflow
+
+Getting your models and tables synced up is a bit tricky. Pay close attention to the following workflow, especially the rake tasks.
+
+```
+# create a new rails app
+rails new my_app -d postgresql
+cd my_app
+
+# create the database
+rake db:create
+
+# REPEAT THESE TASKS FOR EVERY CHANGE TO YOUR DATABASE
+# <<< BEGIN WORKFLOW LOOP >>>
+
+# -- IF YOU NEED A NEW MODEL --
+# auto-generate a new model (AND automatically creates a new migration)
+rails g model Pet name:string
+rails g model Owner name:string
+
+# --- OTHERWISE ---
+
+# if you only need to change fields in an *existing* model,
+# you can just generate a new migration
+rails g migration AddAgeToOwner age:integer
+
+# never try to create a migration file yourself through the file system! it's really hard to get the name right!
+
+# -- EITHER WAY --
+### whether we're creating a new model or updating an existing one, we can manually edit our models and migrations in sublime
+# update associations in model --> this affects model interface
+# update foreign keys in migrations --> this affects database tables
+
+# generate schema for database tables
+rake db:migrate
+
+# <<< END LOOP >>>
+
+# finally, we need some data to play with
+# for now, we'll seed it manually, from the rails console...
+rails c
+> Pet.create(name: "Wowzer")
+> Pet.create(name: "Rufus")
+
+# --- OR ---
```

Comment:
I'm removing the ---OR--- line for now since it doesn't make sense. But I think it might have been a placeholder for talking about using FFaker data. Feel free to change later.